src/execution/execute.ts

```diff
@@ -286,87 +272,321 @@ export function buildExecutionContext(
-    fieldResolver: fieldResolver ?? defaultFieldResolver,
-    typeResolver: typeResolver ?? defaultTypeResolver,
-    subscribeFieldResolver: subscribeFieldResolver ?? defaultFieldResolver,
-    errors: [],
+    fragments,
+    rootType,
+    fieldResolver: options.fieldResolver ?? defaultFieldResolver,
+    typeResolver: options.typeResolver ?? defaultTypeResolver,
+    subscribeFieldResolver:
+      options.subscribeFieldResolver ?? defaultFieldResolver,
   };
+  return { result };
 }
 
 /**
- * Implements the "Executing operations" section of the spec.
+ * Constructs a ExecutableRequest object
  */
-function executeOperation(
-  exeContext: ExecutionContext,
-  operation: OperationDefinitionNode,
-): PromiseOrValue<ObjMap<unknown>> {
-  const rootType = exeContext.schema.getRootType(operation.operation);
-  if (rootType == null) {
-    throw new GraphQLError(
-      `Schema is not configured to execute ${operation.operation} operation.`,
-      { nodes: operation },
-    );
-  }
-
-  const rootFields = collectFields(
-    exeContext.schema,
-    exeContext.fragments,
-    exeContext.variableValues,
-    rootType,
-    operation.selectionSet,
+export function makeExecutableRequest(
+  schema: GraphQLSchema,
+  document: DocumentNode,
+  operationName: Maybe<string>,
+  options: GraphQLExecutionPlanOptions = {},
+): ResultOrGraphQLErrors<ExecutableRequest> {
+  const makeExecutionPlanReturn = makeExecutionPlan(
+    schema,
+    document,
+    operationName,
+    options,
   );
-  const path = undefined;
 
-  const { rootValue } = exeContext;
+  if (makeExecutionPlanReturn.errors !== undefined) {
+    return makeExecutionPlanReturn;
+  }
+  const executionPlan = makeExecutionPlanReturn.result;
 
-  switch (operation.operation) {
+  switch (executionPlan.operation.operation) {
     case OperationTypeNode.QUERY:
-      return executeFields(exeContext, rootType, rootValue, path, rootFields);
+      return { result: new ExecutableQueryRequestImpl(executionPlan) };
     case OperationTypeNode.MUTATION:
-      return executeFieldsSerially(
-        exeContext,
-        rootType,
-        rootValue,
-        path,
-        rootFields,
-      );
+      return { result: new ExecutableMutationRequestImpl(executionPlan) };
     case OperationTypeNode.SUBSCRIPTION:
-      // TODO: deprecate `subscribe` and move all logic here
-      // Temporary solution until we finish merging execute and subscribe together
-      return executeFields(exeContext, rootType, rootValue, path, rootFields);
+      return { result: new ExecutableSubscriptionRequestImpl(executionPlan) };
+  }
+}
+
+class CoercedVariableValues {
+  coercedValues: { [variable: string]: unknown };
+
+  constructor(objMap: { [variable: string]: unknown }) {
+    this.coercedValues = objMap;
+  }
+}
+
+export type ExecutableRequest =
+  | ExecutableQueryRequest
+  | ExecutableMutationRequest
+  | ExecutableSubscriptionRequest;
+
+export interface ExecutableQueryRequest {
```

Comment:
@IvanGoncharov  This is the key comment I believe in my review of this approach. It's more of a question than a comment.

I know I have asked elsewhere to export interfaces so that third parties can customize, add hooks, etc. I am generally happy to see the suggestion adopted, but I have to say that I don't know if the added complexity in this particular instance is so helpful. We currently have a pipeline of schema generation => request parse => request validate => request execute. Hidden in there also as a step is that without "assumeValid", schema validation is automatically performed.

My thinking in overall direction was to break out additional recommended steps, i.e., "validateSchema" would be its own step, request document parsing would be its own step, variable coercion would be its own step, and execute would then operation on the results of the previous stages, such that ExecutionContext would be broken down into its constituent parts, some cacheable (document), some not cacheable/not as cacheable (variables). Perhaps this interface/implementation approach has some additional benefits I am not appreciating?

Reply:
> Hidden in there also as a step is that without "assumeValid", schema validation is automatically performed.

It's not hidden, it's "optional" you can perform it explicitly if you want.
All GraphQL servers I saw just call `validateSchema` and even our own `graphql` function does that, for example, https://github.com/graphql/express-graphql/blob/f4414b44996f25a5328e523d9a4b213fd1d70b16/src/index.ts#L268
But I agree it's a bad API design so I plan to work on the `GraphQLExecutableSchema` class to make it more explicit from an API standpoint.

> I know I have asked elsewhere to export interfaces so that third parties can customize, add hooks, etc. I am generally happy to see the suggestion adopted, but I have to say that I don't know if the added complexity in this particular instance is so helpful.

I still have the same position about hooks, this PR adds **interfaces** that can be used by external libraries to customize executor.
Different here is that we don't have any single line of code that can run 3rd-party `ExecutableRequest`, so it's not a hook.

I'm for adding configuration/interfaces if the below rules are satisfied:
1. Use case for which we add it is fully spec-compliant.
2. We can't incorporate this use case in graphql-js. For example, there is a "one size fits all" solution.

Custom GraphQL executors satisfy both rules, JIT is a valid use case but JIT is not a universal one (not everyone trusts the code that `eval`) it also applies some limitations: https://github.com/zalando-incubator/graphql-jit#differences-to-graphql-js
Even JIT aside, custom execute engines purposely built for complex scenarios like federation, and the stitching will always outperform generic ones even with all necessary hooks provided (hooks add their own cost and prevent certain JIT/compilation optimizations).
So it's a valid use case for adding interfaces.

This PR basically implements the https://github.com/graphql/graphql-js/issues/3314 proposal and defines clean interfaces that can be used by GraphQL servers to support specialized GraphQL executors.

> I have to say that I don't know if the added complexity in this particular instance is so helpful.

Just to be clear, are you referring to interfaces specifically or your comment is about the entire approach of using class instances vs having functions accept query plan as an argument?

Reply:
I don't think we need an ExecutableSchema class. I think if validateSchema is not called, the schema should not be validated, plain and simple.

In terms of your last question -- I am not sure I understand the distinction. I commented below in terms of relationship to #3314.

Basically, if the goal is to separate out coercion, and storing result of document analysis into separate step, I think that can be done as separate exported functions, like analyzeDocument and coerceVariables, and then the user can cache them. We actually don't have to deprecate execute, we can still include it for those who want it but just be explicit about which exported pipeline steps it performs: analyzeDocument, coerceVariables, getRootType, executeOperation...